ros2_foxglove_bridge/src/ros2_foxglove_bridge.cpp

```diff
@@ -1,24 +1,367 @@
+    _updateTimer = this->create_wall_timer(
+      nextUpdateMs, std::bind(&FoxgloveBridge::updateAdvertisedTopics, this));
   }
 
 private:
+  struct PairHash {
+    template <class T1, class T2>
+    std::size_t operator()(const std::pair<T1, T2>& pair) const {
+      return std::hash<T1>()(pair.first) ^ std::hash<T2>()(pair.second);
+    }
+  };
+
+  foxglove::Server<foxglove::WebSocketNoTls> _server;
+  foxglove::MessageDefinitionCache _messageDefinitionCache;
+  std::unordered_map<TopicAndDatatype, foxglove::Channel, PairHash> _advertisedTopics;
+  std::unordered_map<foxglove::ChannelId, TopicAndDatatype> _channelToTopicAndDatatype;
+  std::unordered_map<foxglove::ChannelId, std::shared_ptr<rclcpp::GenericSubscription>>
+    _subscriptions;
+  std::mutex _subscriptionsMutex;
+  rclcpp::node_interfaces::OnSetParametersCallbackHandle::SharedPtr _parametersUpdateHandle;
+  rclcpp::TimerBase::SharedPtr _updateTimer;
+  size_t _maxUpdateMs = size_t(DEFAULT_MAX_UPDATE_MS);
+  size_t _updateCount = 0;
+
+  rcl_interfaces::msg::SetParametersResult parametersHandler(
+    const std::vector<rclcpp::Parameter>& parameters) {
+    for (const auto& parameter : parameters) {
+      if (parameter.get_name() == "max_update_ms") {
+        _maxUpdateMs = size_t(parameter.as_int());
+      }
+    }
+
+    rcl_interfaces::msg::SetParametersResult result;
+    result.successful = true;
+    result.reason = "success";
+    return result;
+  }
+
+  void subscribeHandler(foxglove::ChannelId channelId) {
+    std::lock_guard<std::mutex> lock(_subscriptionsMutex);
+    if (_subscriptions.find(channelId) != _subscriptions.end()) {
+      // Subscription already exists
+      RCLCPP_DEBUG(this->get_logger(), "Subscription already exists for channel %d", channelId);
+      return;
+    }
+
+    auto it = _channelToTopicAndDatatype.find(channelId);
+    if (it == _channelToTopicAndDatatype.end()) {
+      RCLCPP_WARN(this->get_logger(), "Received subscribe request for unknown channel %d",
+                  channelId);
+      return;
+    }
+    auto& topicAndDatatype = it->second;
+    auto topic = topicAndDatatype.first;
+    auto datatype = topicAndDatatype.second;
+    auto it2 = _advertisedTopics.find(topicAndDatatype);
+    if (it2 == _advertisedTopics.end()) {
+      RCLCPP_ERROR(this->get_logger(), "Channel %d for topic \"%s\" (%s) is not advertised",
+                   channelId, topic.c_str(), datatype.c_str());
+      return;
+    }
+    auto& channel = it2->second;
+
+    rclcpp::SubscriptionEventCallbacks eventCallbacks;
+    eventCallbacks.incompatible_qos_callback = [&](const rclcpp::QOSRequestedIncompatibleQoSInfo&) {
+      RCLCPP_ERROR(this->get_logger(), "Incompatible subscriber QoS settings for topic \"%s\" (%s)",
+                   topic.c_str(), datatype.c_str());
+    };
+
+    rclcpp::SubscriptionOptions subscriptionOptions;
+    subscriptionOptions.event_callbacks = eventCallbacks;
```

Comment:
We should also add a dedicated [callback_group](https://docs.ros2.org/galactic/api/rclcpp/structrclcpp_1_1SubscriptionOptionsBase.html#ab3998ee55c81d8035a0bd4b574200218) for each subscription. Otherwise all subscription share the default callback group and there is little multi threading (despite the multi threaded executor)

Reply:
https://github.com/foxglove/ros-foxglove-bridge/issues/16